lib/krb5/store_stdio.c

```diff
@@ -0,0 +1,231 @@
+{
+    char *cbuf = (char *)data;
+    ssize_t count;
+    size_t rem = size;
+
+    /* similar pattern to net_read() to support pipes */
+    while (rem > 0) {
+	count = fread(cbuf, 1, rem, F(sp));
+	if (count < 0) {
+	    if (errno == EINTR)
+		continue;
+	    else
+		return count;
+	} else if (count == 0) {
+	    return count;
+	}
+	cbuf += count;
+	rem -= count;
+    }
+    return size;
+}
+
+static ssize_t
+stdio_store(krb5_storage * sp, const void *data, size_t size)
+{
+    const char *cbuf = (const char *)data;
+    ssize_t count;
+    size_t rem = size;
+
+    /* similar pattern to net_write() to support pipes */
+    while (rem > 0) {
+	count = fwrite(cbuf, 1, rem, F(sp));
+	if (count < 0) {
+	    if (errno == EINTR)
+		continue;
+	    else
+		return count;
+	}
+	cbuf += count;
+	rem -= count;
+    }
+    return size;
+}
+
+static off_t
+stdio_seek(krb5_storage * sp, off_t offset, int whence)
+{
+    int save_errno = errno;
+
+    if (fseeko(F(sp), offset, whence) != 0)
+        return -1;
+    errno = save_errno;
+    return ftello(F(sp));
+}
+
+static int
+stdio_trunc(krb5_storage * sp, off_t offset)
+{
+    int save_errno = errno;
+
+    if (fflush(F(sp)) == EOF)
+        return errno;
+    if (ftruncate(fileno(F(sp)), offset) == -1)
+	return errno;
+    if (fseeko(F(sp), offset, SEEK_SET) == -1)
+        return errno;
+    errno = save_errno;
+    return 0;
+}
+
+static int
+stdio_sync(krb5_storage * sp)
+{
+    if (fflush(F(sp)) == EOF)
+	return errno;
+    if (fsync(fileno(F(sp))) == -1)
+	return errno;
+    return 0;
+}
+
+static void
+stdio_free(krb5_storage * sp)
+{
+    int save_errno = errno;
+    if (fclose(F(sp)) == 0)
+        errno = save_errno;
+}
+
+/**
+ * Open a krb5_storage using stdio for buffering.
+ *
+ * @return A krb5_storage on success, or NULL on out of memory error.
+ *
+ * @ingroup krb5_storage
+ *
+ * @sa krb5_storage_emem()
+ * @sa krb5_storage_from_fd()
+ * @sa krb5_storage_from_mem()
+ * @sa krb5_storage_from_readonly_mem()
+ * @sa krb5_storage_from_data()
+ * @sa krb5_storage_from_socket()
+ */
+
+KRB5_LIB_FUNCTION krb5_storage * KRB5_LIB_CALL
+krb5_storage_stdio_from_fd(int fd_in)
+{
+    krb5_storage *sp;
+    const char *mode = "r";
+    off_t off;
+    FILE *f;
+    long fl;
+    int saved_errno = errno;
+    int fd;
+
+    fl = fcntl(fd_in, F_GETFL);
+    if (fl == -1)
+        return NULL;
+
+    off = lseek(fd_in, 0, SEEK_CUR);
+    if (off == -1)
+        return NULL;
+
+#ifdef _MSC_VER
+    /*
+     * This function used to try to pass the input to
+     * _get_osfhandle() to test if the value is a HANDLE
+     * but this doesn't work because doing so throws an
+     * exception that will result in Watson being triggered
+     * to file a Windows Error Report.
+     */
+    fd = _dup(fd_in);
+#else
+    fd = dup(fd_in);
+#endif
+
+    if (fd < 0)
+        return NULL;
+
+    if ((fl & O_RDWR))
+        mode = "r+";
+    else if ((fl & O_WRONLY))
+        mode = "w";
```

Comment:
The O_RDONLY, O_WRONLY and O_RDWR flags are not required to be independent bits!  Instead switch on (fl & O_ACCMODE)

http://pubs.opengroup.org/onlinepubs/009695399/functions/fcntl.html